build and push docker images on git tag also, split test, add npm audit

misiek-sf opened the pull request:
in this state, You can use it like that:
```yaml
workflows:
  version: 2
  build_test_and_deploy:
    jobs:
      - ms/setup-env:
          context: microservices
          filters:
              tags:
                only: /^v.*/
      - ms/build-image:
          context: microservices
          requires:
            - ms/setup-env
          filters:
              tags:
                only: /^v.*/
      - ms/push-image:
          context: microservices
          requires:
            - ms/build-image
            - ms/run-unit-tests
            # - ms/run-integration-tests
            - ms/run-lint
          filters:
              tags:
                only: /^v.*/
      - ms/run-unit-tests:
          context: microservices
          requires:
            - ms/build-image
          filters:
              tags:
                only: /^v.*/
      # NOTE: turned off till vpn issues are resolved
      # - ms/run-integration-tests:
      #     context: microservices
      #     requires:
      #       - ms/build-image
      #     filters:
      #         tags:
      #           only: /^v.*/
      - ms/run-lint:
          context: microservices
          requires:
            - ms/build-image
          filters:
              tags:
                only: /^v.*/
      - hold:
          type: approval
          requires:
            - ms/push-image
          filters:
            branches:
              only:
                - prod
      - ms/deploy-svc:
          context: microservices
          requires:
            - hold
            - ms/push-image
          filters:
            *ms-deployable-branches-filter
```

on git tag, there will be an image in ECR tagged exactly as git tag
also there will be an image in ECR tagged with branch name


example "view" in circleci:
![image](https://user-images.githubusercontent.com/45633029/50597982-93983b00-0ea9-11e9-903a-286f0765165d.png)

and with integration tests turned off:
![image](https://user-images.githubusercontent.com/45633029/50598013-ab6fbf00-0ea9-11e9-9f9b-3faf16bc2d0d.png)


----
update with `npm audit`:
![image](https://user-images.githubusercontent.com/45633029/50643487-96a92f00-0f6e-11e9-803d-911edb5fa1f3.png)


```yaml
workflows:
  version: 2
  build_test_and_deploy:
    jobs:
      - ms/setup-env:
          context: microservices
          filters:
              tags:
                only: /^v.*/
      - ms/build-image:
          context: microservices
          requires:
            - ms/setup-env
          filters:
              tags:
                only: /^v.*/
      - ms/push-image:
          context: microservices
          requires:
            - ms/build-image
            - ms/run-unit-tests
            # - ms/run-integration-tests
            - ms/run-lint
            - ms/npm-audit
          filters:
              tags:
                only: /^v.*/
      - ms/run-unit-tests:
          context: microservices
          requires:
            - ms/build-image
          filters:
              tags:
                only: /^v.*/
      # NOTE: turned off till vpn issues are resolved
      # - ms/run-integration-tests:
      #     context: microservices
      #     requires:
      #       - ms/build-image
      #     filters:
      #         tags:
      #           only: /^v.*/
      - ms/run-lint:
          context: microservices
          requires:
            - ms/build-image
          filters:
              tags:
                only: /^v.*/
      - ms/npm-audit:
          filters:
              tags:
                only: /^v.*/
      - hold:
          type: approval
          requires:
            - ms/push-image
          filters:
            branches:
              only:
                - prod
      - ms/deploy-svc:
          context: microservices
          requires:
            - hold
            - ms/push-image
          filters:
            *ms-deployable-branches-filter
```

misiek-sf committed:
build and push docker images on git tag also

wprater commented on microservices.yml:
```diff
@@ -156,7 +159,16 @@ jobs:
       - print-diagnostics
       - fetch-environment-variables-from-s3
 
-  # Step 2: Verify that tests are passing
```
Why would we want to push an image with potentially failing tests?

misiek-sf replied:
yeah... we can use some complicated logic (e.g. `docker save... presist_to_workspace... docker load...`), but we can also remove images for failed tests (`when: on_fail`)... 

or... we can try with `setup_remote_docker` step
(also remember that "for now" tests are turned off... so we can get back to this when we will be turning them on)

what do u think?

tomasz-sf replied:
I see no purpose pushing builds with failing tests to ECR. Maybe just save to workspace?

misiek-sf replied:
i;m all on not pushing failing builds. 
but how do you think we should approach this; as each "step" is run on different executor so built image needs to be "transferred" to next executor, if we don't use some repo...

any ideas?

misiek-sf replied:
`setup_remote_docker` step is not compatible with machine executor... 

> Build-agent version 0.1.1250-22bf9f5d (2018-12-12T11:32:15+0000)
> Configuration errors: 1 error occurred:
> 
> * In step 6 (setup_remote_docker): Step isn't compatible with machine executor.  You can use Docker without a special step.

so `persist` and `attach`?

misiek-sf replied:
`persist && attach` "example": https://circleci.com/workflow-run/c818a3d5-743f-4f7c-a898-aed7fb5d15ef

misiek-sf replied:
and the "example" of `setup` -> `build` -> `test` (failing) -> `push` 
https://circleci.com/workflow-run/595d6e32-0c11-4d08-bf6f-6f8d0749c337

wprater replied:
sounds good to leverage the workspace and docker for this.  do remember that the `machine` executor is required for our vpn access only, so we cannot remove it or use `setup_remote_docker`.

misiek-sf replied:
I've updated this PR in the way that now the "push to ECR" step can require the tests to pass

an example of this working is in https://github.com/MGMResorts/devices/pull/416

wprater commented on version.json:
```diff
@@ -1,4 +1,4 @@
 {
   "name": "mgmorbs/microservices",
-  "version": "0.3.1"
+  "version": "0.4.0"
```
did you actually publish a dev version of this orb yet?  we should do this inside of circleCI and be sure to run `circleci config validate` to ensure the config is valid.

misiek-sf replied:
no, i did not; i wanted to get team opinion before
config validate is automatic on this repo (one of the checks of PR) and also on a tag in this repo we will get auto-publishing (take a look on `.circleci/config.yaml`)

(also i'v done `circleci orb validate.. && circleci orb process...` on my machine)

wprater replied:
more just curious, but we should come to a process on this.  I think it's safe to always publish to dev with `npm version` or in Circle, since dev is just arbitrary version tags and production orbs are all SemVer.

misiek-sf replied:
what I've done:
- on each `branch` push there will be published (after validating of course) orb with tag `@dev:BRANCHNAME`
- on `git tag` there will be published a new "public" version of this orb (with the version from `version.json`)

so you can use a development version of this orb in your `.circleci/config.yaml` e.g.:
```yaml
orbs:
  # Repo: https://github.com/MGMDV-Orbs/microservices
  # Latest Version: https://circleci.com/orbs/registry/orb/mgmorbs/microservices
  ms: mgmorbs/microservices@dev:f/SRV-188
```

this repo now needs one approve from reviewers and to `validate` phase to pass for every PR

misiek-sf committed:
publish dev version of orb

on each push to this repo circleci will publish orb
with development tag `mgmorbs/microservices@dev:BRANCH`

misiek-sf committed:
separate build and push step, add tests inside container

also publish test results (only eslint for now)

misiek-sf committed:
fix step descriptions

misiek-sf commented on version.json:
```diff
@@ -1,4 +1,4 @@
 {
   "name": "mgmorbs/microservices",
-  "version": "0.3.1"
+  "version": "1.0.0"
```
I've updated major version as steps are incompatible with old ones

afaq-sf commented on microservices.yml:
```diff
@@ -77,30 +98,35 @@ commands:
+      - run:
+          name: Run lint in docker container
+          command: |
+            docker run -v /home/circleci/project/reports:/usr/src/app/reports -it ${TAG_COMMIT} npm run ci:lint
```
@misiek-sf Does `npm run ci:lint` script exist in all repos yet?

misiek-sf replied:
yep (gotten there with `linting and stuff`)

codin-sf commented on microservices.yml:
```diff
@@ -58,16 +59,36 @@ commands:
+            docker build -t $TAG_LATEST -t $TAG_BRANCH -t $TAG_COMMIT -f Dockerfile .
+            [ ! -z $TAG_GITTAG ] && docker tag $TAG_COMMIT $TAG_GITTAG || true
+      - run:
+          name: Save docker image
```
I don't think this matters since it is a VM but do we need a "cleanup" task to delete what is in the images directory?

misiek-sf replied:
afaik the VMs that are used in CircleCI get trashed after use (you can "restart tests with ssh" and then VM is kept for up to 2h)

sooo i wouldn't bother with cleanup

codin-sf commented on microservices.yml:
```diff
@@ -77,30 +98,35 @@ commands:
-
-  run-docker-compose-tests:
-    description: Runs docker compose up tests with proper exit code verification to fail build
+            docker run --env LOG_TO_CONSOLE=true --env LOG_TO_FILE=true --env LOGGING_LEVEL=info --env-file common.env --env-file tables.env --env-file assa.env --env-file dorma.env -it ${TAG_COMMIT} npm run test:unit
```
the env-file definitions are a bit questionable. We have services with many (an example is the rooms has common, logss3, logcloudwatch, snt, rover). Do we need to specify all possible options in this line?

misiek-sf replied:
I'll resolve this conversation to keep it in one thread

codin-sf commented on microservices.yml:
```diff
@@ -77,30 +98,35 @@ commands:
+          name: Run integration tests in docker container
           command: |
-            docker-compose up --exit-code-from test test
+            docker run --env LOG_TO_CONSOLE=true --env LOG_TO_FILE=true --env LOGGING_LEVEL=info --env-file common.env --env-file tables.env --env-file assa.env --env-file dorma.env -it ${TAG_COMMIT} npm run test:integration
```
the env-file definitions are a bit questionable. We have services with many (an example is the rooms has common, logss3, logcloudwatch, snt, rover). Do we need to specify all possible options in this line?

bartek-sf replied:
should be agnostic, can we source variables from `BASH_ENV` into container?

misiek-sf replied:
yeah, this is my main concern... sourcing `BASH_ENV` is hacky at least...

my proposition is to use `Makefile` in each repo and run `make ci:test:unit`/`make ci:test:integration`. that way we can do lots of stuff the same in each repo but also we can have some exception per repo and still use the common way of running on Circle, also we can "source" `BASH_ENV` inside `Makefile`

what do you think guys @codin-sf @bartek-sf @afaq-sf @tomasz-sf @patrick-sf

codin-sf replied:
I like this idea

👍

codin-sf reviewed (changes requested):
I have left some comments to gain feedback on. Also, I think it makes sense that we add the tag version dynamically to (possibly?) the package.json version so when we hit the /healthcheck endpoint we know for certain what we are looking at. Thoughts?

codin-sf commented:
In the future, let's have smaller PRs with succinct goals. This way we can add items as approved rather than blocking an entire group of edits because there are so many topics.

In addition to the steps above. Do we think we should add an ` npm audit ` task (commented out for now until we fix issues)?

bartek-sf commented on microservices.yml:
```diff
@@ -58,16 +59,36 @@ commands:
           command: |
             echo 'export TAG_LATEST="${ECR_APP_IMAGE}:latest"' >> $BASH_ENV
-            echo 'export TAG_CI="${ECR_APP_IMAGE}:BUILD_${CIRCLE_BUILD_NUM}"' >> $BASH_ENV
+            echo 'export TAG_BRANCH="${ECR_APP_IMAGE}:BRANCH_${CIRCLE_BRANCH/\//_}"' >> $BASH_ENV
```
is it required `/\//_`?

misiek-sf replied:
yep... you can't use a tag with `/` I'm replacing `/` to `_` but I'm open to suggestion on that 😞

bartek-sf replied:
sure now it is clear, looks good.

misiek-sf commented:
> In the future, let's have smaller PRs with succinct goals. This way we can add items as approved rather than blocking an entire group of edits because there are so many topics.
> 
> In addition to the steps above. Do we think we should add an `npm audit` task (commented out for now until we fix issues)?

good catch.. will do that

misiek-sf committed:
add npm audit job

misiek-sf commented:
I've added `npm-audit` step; for now, it breaks the build for vulnerabilities of the `critical` level

should be changed do high or moderate after fixing our code

_updated description_

misiek-sf commented:
note to self: before merge update `README.md`

misiek-sf commented:
> In the future, let's have smaller PRs with succinct goals. This way we can add items as approved rather than blocking an entire group of edits because there are so many topics.

totally agree on that, I've been lured to this by @wprater 😉 comment of pushing to ECR after failed tests; what i should have done is make another PR for this

misiek-sf committed:
use make to run unit and integration tests

misiek-sf committed:
lint outside of built image

using make

misiek-sf commented:
> I have left some comments to gain feedback on. Also, I think it makes sense that we add the tag version dynamically to (possibly?) the package.json version so when we hit the /healthcheck endpoint we know for certain what we are looking at. Thoughts?

@codin-sf i was (re)searching for this, and from my knowledge, it is almost impossible without some hacks and stuff, but the reasoning for it was great, and we should return to "something" "like that" later.

wprater commented:
> > I have left some comments to gain feedback on. Also, I think it makes sense that we add the tag version dynamically to (possibly?) the package.json version so when we hit the /healthcheck endpoint we know for certain what we are looking at. Thoughts?
> 
> @codin-sf i was (re)searching for this, and from my knowledge, it is almost impossible without some hacks and stuff, but the reasoning for it was great, and we should return to "something" "like that" later.

Not sure I understand what is being proposed here?  Returning the version from the `/healthcheck` endpoint??

misiek-sf commented:
> > > I have left some comments to gain feedback on. Also, I think it makes sense that we add the tag version dynamically to (possibly?) the package.json version so when we hit the /healthcheck endpoint we know for certain what we are looking at. Thoughts?
> > 
> > 
> > @codin-sf i was (re)searching for this, and from my knowledge, it is almost impossible without some hacks and stuff, but the reasoning for it was great, and we should return to "something" "like that" later.
> 
> Not sure I understand what is being proposed here? Returning the version from the `/healthcheck` endpoint??

i understand it like that:
1. build docker image with tag "9.0.0-alfa-super-fun-version-which-never-fails"
1. start service
1. `/healthcheck` returns `version: "9.0.0-alfa-super-fun-version-which-never-fails"`

codin-sf commented:
> > > > I have left some comments to gain feedback on. Also, I think it makes sense that we add the tag version dynamically to (possibly?) the package.json version so when we hit the /healthcheck endpoint we know for certain what we are looking at. Thoughts?
> > > 
> > > 
> > > @codin-sf i was (re)searching for this, and from my knowledge, it is almost impossible without some hacks and stuff, but the reasoning for it was great, and we should return to "something" "like that" later.
> > 
> > 
> > Not sure I understand what is being proposed here? Returning the version from the `/healthcheck` endpoint??
> 
> i understand it like that:
> 
> 1. build docker image with tag "9.0.0-alfa-super-fun-version-which-never-fails"
> 2. start service
> 3. `/healthcheck` returns `version: "9.0.0-alfa-super-fun-version-which-never-fails"`

Yes that is what I had in mind. it would be more like : 

`version: "2.0.0"`
`build: "9.0.0-alfa-super-fun-version-which-never-fails"`

 I don't know if this even makes sense. If we properly update our `npm version` and track what we are doing perhaps we are ok.

wprater commented:
> i understand it like that:
> 
> 1. build docker image with tag "9.0.0-alfa-super-fun-version-which-never-fails"
> 2. start service
> 3. `/healthcheck` returns `version: "9.0.0-alfa-super-fun-version-which-never-fails"`

Got it!  In that case, why not just use the `version` field as it's been designed?  SemVer allows for pre-release tags and the only time we'd want those tags as you described them was for `prerelease` anyhow.

https://semver.org/#spec-item-9

i.e. run a command like `npm version prerelease alfa-super-fun-version-which-never-fails` to append to the current version.  You can use the npm script hooks `preversion` and `postversion` to publish the orb, etc. :)

misiek-sf committed:
change building of docker image to use makefile

misiek-sf committed:
push image by make

misiek-sf committed:
fix description in attach-built-docker-image

misiek-sf committed:
update readme

misiek-sf committed:
remove setting full image name inside orb

misiek-sf committed:
remove all-in-one step

codin-sf reviewed (approved):
Since @wprater and @afaq-sf have been so heavily involved with the orbs I would request an approval from one of them too. It appears this is ready to go though.

afaq-sf commented on README.md:
```diff
@@ -7,53 +7,78 @@ https://circleci.com/orbs/registry/orb/mgmorbs/microservices
+          context: microservices
+          filters:
+              tags:
+                only: /^v.*/
```
@misiek-sf Can you please elaborate on why we're only building on tags and not on every commit of filtered branches?

misiek-sf replied:
@afaq-sf this is only public readme example, "we" are doing it just like you suggested on the call 😉 

so there is no need to do new PR

misiek-sf replied:
i can update the readme if required
